### ISSUE TITLE
Provide "higher contrast" versions of themes to comply with WCAG?

### ISSUE THREAD

**Issue by maelle** [alecthomas/chroma#353]
Thank you for creating&maintaining Chroma that I've been excited to use via Hugo! :sparkles: 

Disclaimer: I am not an accessibility expert!

**What problem does this feature solve?**

Some Chroma styles, e.g. friendly or vim, have colors that do not comply with [WCAG 2.0 guidelines on color contrast](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html). For instance in friendly, comments are too light on the background.

I used [this online tool](https://color.a11y.com/) to check websites of mine. One can also check https://xyproto.github.io/splash/docs/longer/all.html with it.

**What feature do you propose?**

I'm wondering whether there could be a few high contrast themes, or a high contrast version for all themes.

## High contrast themes

I only know of [a11y](https://github.com/ericwbailey/a11y-syntax-highlighting) as a theme that's advertised as accessible. It doesn't have a Chroma stylesheet. I tried creating one myself but I'm limited by my understanding of all Chroma classes to do a good job.

Codepen has its own [high contrast themes](https://blog.codepen.io/2019/09/09/new-high-contrast-syntax-highlighting-themes/), that are probably copyrighted (I am not a lawyer either, but they are based on Codepen official colors).

## Amending themes

Max Chadwick [created a fork of some Pygments styles](https://maxchadwick.xyz/blog/syntax-highlighting-and-color-contrast-accessibility), first by hand, then using a Ruby script, as explained in the blog post. I myself darkened some of friendly colors by hand.

**Comment by alecthomas** [comment]
chroma includes an importer script for pygments colour schemes, have you tried it?

**Comment by maelle** [comment]
Oh, nice, I can use that to import https://github.com/ericwbailey/a11y-syntax-highlighting and  https://github.com/mpchadwick/pygments-high-contrast-stylesheets/ then, thank you! I had missed that.

But I was also wondering whether high contrast themes could become "built-in"?

**Comment by alecthomas** [comment]
For sure. Send me a PR and I'll be happy to add them.

**Comment by maelle** [comment]
Cool! I'll see what I can do. Three questions

* The repo by @mpchadwick is unlicensed, how do I properly acknowledge the origin of the styles?

* In practice, I "just" need to add one .go file per style, correct?

* What should be the naming scheme? Adding "-hc" to the end of the original stylename?

**Comment by maelle** [comment]
Another question, is the tool to use https://github.com/alecthomas/chroma/blob/master/_tools/pygments2chroma.py?

**Comment by derat** [comment]
In case it helps anyone else, here's a quick function to perform ad-hoc brightness adjustments to the colors in a style:

```go
// clampBrightness returns a copy of orig with colors adjusted to be in [min, max] in the range [0, 1].
func clampBrightness(orig *chroma.Style, min, max float64) (*chroma.Style, error) {
    b := chroma.NewStyleBuilder("")
    for _, tt := range orig.Types() {
        se := orig.Get(tt)
        if br := se.Colour.Brightness(); br < min {
            se.Colour = se.Colour.Brighten(min - br)
        } else if br > max {
            se.Colour = se.Colour.Brighten(max - br)
        }
        b.Add(tt, se.String())
    }
    return b.Build()
}
```

I'm not bothering to check contrast against a particular background, but this was enough for me to adjust some styles to pass the Chrome Lighthouse accessibility audits (by calling with `0.85, 1` for a dark background and `0, 0.1` for a light background). Using `Style.Builder` might be more efficient; I abstracted this from some code that was doing more stuff with styles.

(Thanks for Chroma! It's nice that stuff like this can be done without needing to modify the library.)

**Comment by alecthomas** [comment]
I'm not sure why I closed this as it didn't get resolved really. @derat would you mind contributing a PR adding this function, I think it would be generally useful.

I'll also have a ponder how this functionality could be used by consumers of Chroma who aren't writing Go.

**Comment by derat** [comment]
> I'm not sure why I closed this as it didn't get resolved really. @derat would you mind contributing a PR adding this function, I think it would be generally useful.

Sure, can do. Would you prefer something like:

```go
func (c Colour) ClampBrightness(min, max float64) Colour
func (s *Style) ClampBrightness(min, max float64) *Style
```

? If so, should the method on `Style` also clamp the `Background` and `Border` fields, or only `Colour`?

And just to mention it, the clamping logic in the code that I attached is totally bogus, in that the resulting brightness isn't necessarily going to be in the requested range. For example, `(192, 192, 192)` has a brightness close to 0.75 per the `Brightness` function, but calling `Brighten(0.05)` like my code does will yield `(195, 195, 195)`, which only has a brightness of 0.7647 (rather than 0.8) per `Brightness`. If there's a `ClampBrightness` method on `Colour`, it'd probably be better if it returned `(204, 204, 204)` in this case.

**Comment by alecthomas** [comment]
Good question and I'm not sure. Thinking about it, maybe a better idea is to add a `func (s *StyleBuilder) Transform(transform func(StyleEntry)StyleEntry) *StyleBuilder` method, which would mutate the `StyleBuilder` by applying `transform` to each entry.

As to whether the background and border colours should be clamped, I'm not sure. I'm really not familiar with accessibility.

**Comment by derat** [comment]
The semantics of that `Transform` function feel a bit tricky to me. It's essentially a no-op if called on a receiver that was created using `NewStyleBuilder`, right? It seems like it'd only be useful after a `Style.Builder` call.

Regarding which colors to change, I think that the main A11Y concern here is contrast. Clamping both the foreground and background colors' brightnesses within a given range seems like it'd actually reduce contrast. Borders look like they're rarely used in the checked-in styles. 9 styles set a red border for `Error` tokens, while `manni` and `rainbow_dash` (both light-background styles) set darker red/green borders around `GenericDeleted` and `GenericInserted` (which have light red/green backgrounds). I suspect that leaving border colors unchanged would be best for readability.

Also, are there any styles that use inverted colors in some cases (e.g. a dark-foreground, light-background style that uses light text on a dark background for errors)? If so, those would also end up with reduced contrast if the user tries to darken the style's colors.

I'll make a PR that adds `ClampBrightness` functions for both `Colour` and `Style`, but given the subtleties around how styles should be adjusted, maybe it'd be best if only the `Colour` function is added.
